crates/polars-stream/src/nodes/parquet_source.rs

```diff
@@ -0,0 +1,1734 @@
+            .step_by(cols_per_task)
+            .map(move |offset| {
+                let row_group_data = row_group_data_2.clone();
+                let slice_range = slice_range_2.clone();
+                let projected_arrow_fields = projected_arrow_fields.clone();
+
+                async move {
+                    (offset
+                        ..offset
+                            .saturating_add(cols_per_task)
+                            .min(projected_arrow_fields.len()))
+                        .map(|i| {
+                            let arrow_field = projected_arrow_fields[i].clone();
+
+                            let columns_to_deserialize = row_group_data
+                                .row_group_metadata
+                                .columns()
+                                .iter()
+                                .filter(|col_md| {
+                                    col_md.descriptor().path_in_schema[0] == arrow_field.name
+                                })
+                                .map(|col_md| {
+                                    let (offset, len) = col_md.byte_range();
+                                    let offset = offset as usize;
+                                    let len = len as usize;
+
+                                    (
+                                        col_md,
+                                        row_group_data
+                                            .byte_source
+                                            .get_range(offset..offset + len)
+                                            // Safety: The byte source variant does not perform any async calls.
+                                            .now_or_never()
+                                            .unwrap()
+                                            .unwrap(),
+                                    )
+                                })
+                                .collect::<Vec<_>>();
+
+                            assert!(
+                                slice_range.end <= row_group_data.row_group_metadata.num_rows()
+                            );
+
+                            let array = polars_io::prelude::_internal::to_deserializer(
+                                columns_to_deserialize,
+                                arrow_field.clone(),
+                                Some(polars_parquet::read::Filter::Range(slice_range.clone())),
+                            )?;
+
+                            let series = Series::try_from((&arrow_field, array))?;
+
+                            // TODO: Also load in the metadata.
+
+                            PolarsResult::Ok(series)
+                        })
+                        .collect::<PolarsResult<Vec<_>>>()
+                }
+            });
+
+        if self.use_par_decode && decode_fut_iter.len() > 1 {
+            for handle in decode_fut_iter.map(|fut| {
+                async_executor::AbortOnDropHandle::new(async_executor::spawn(
```

Comment:
@orlp one more place where I spawn - this is for decoding the columns within a row group in parallel. I think this makes sense for very wide tables, but from testing the performance was identical for 1M row groups with 50 columns. I've currently tuned it send a minimum of `const VALUES_PER_THREAD: usize = 8_388_608`  per thread

Reply:
Can you maybe add a comment explaining how/why that value was derived?

Reply:
Added 👍